File: clients/go/examples/test_udp_commands.go
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"time"

	"github.com/blazecache/client"
)

func main() {
	serverAddr := flag.String("server", "127.0.0.1:6793", "UDP server address")
	flag.Parse()

	fmt.Printf("=== Testing Go UDP Client ===\n")
	fmt.Printf("Server: %s\n\n", *serverAddr)

	// Create UDP client
	udpClient, err := blazecache.NewUDPClient(*serverAddr)
	if err != nil {
		log.Fatalf("Failed to create UDP client: %v", err)
	}
	defer udpClient.Close()

	// Test 1: PING
	fmt.Println("1. Testing PING...")
	start := time.Now()
	err = udpClient.Ping()
	elapsed := time.Since(start)
	if err != nil {
		fmt.Printf("   ✗ PING failed: %v\n", err)
	} else {
		fmt.Printf("   ✓ PING successful (took %v)\n", elapsed)
	}
	fmt.Println()

	// Test 2: SET (small value)
	fmt.Println("2. Testing SET (small value)...")
	testKey := "test-key-go"
	testValue := []byte("test-value-go")
	start = time.Now()
	err = udpClient.Set(testKey, testValue, 3600)
	elapsed = time.Since(start)
	if err != nil {
		fmt.Printf("   ✗ SET failed: %v\n", err)
	} else {
		fmt.Printf("   ✓ SET successful (took %v)\n", elapsed)
	}
	fmt.Println()

	// Test 3: GET
	fmt.Println("3. Testing GET...")
	start = time.Now()
	retrievedValue, err := udpClient.Get(testKey)
	elapsed = time.Since(start)
	if err != nil {
		fmt.Printf("   ✗ GET failed: %v\n", err)
	} else {
		if string(retrievedValue) == string(testValue) {
			fmt.Printf("   ✓ GET successful (took %v)\n", elapsed)
			fmt.Printf("   ✓ Value matches: %s\n", string(retrievedValue))
		} else {
			fmt.Printf("   ✗ GET value mismatch!\n")
			fmt.Printf("   Expected: %s\n", string(testValue))
			fmt.Printf("   Got:      %s\n", string(retrievedValue))
		}
	}
	fmt.Println()

	// Test 4: SET (large value - should trigger fragmentation)
	fmt.Println("4. Testing SET (large value - fragmentation)...")
	largeKey := "test-large-key-go"
	largeValue := make([]byte, 5000) // Larger than MAX_DATAGRAM
	for i := range largeValue {
		largeValue[i] = byte(i % 256)
	}
	start = time.Now()
	err = udpClient.Set(largeKey, largeValue, 3600)
	elapsed = time.Since(start)
	if err != nil {
		fmt.Printf("   ✗ SET (large) failed: %v\n", err)
	} else {
		fmt.Printf("   ✓ SET (large) successful (took %v)\n", elapsed)
		fmt.Printf("   ✓ Sent %d bytes in fragments\n", len(largeValue))
	}
	fmt.Println()

	// Test 5: GET (large value - should trigger reassembly)
	fmt.Println("5. Testing GET (large value - reassembly)...")
	start = time.Now()
	retrievedLargeValue, err := udpClient.Get(largeKey)
	elapsed = time.Since(start)
	if err != nil {
		fmt.Printf("   ✗ GET (large) failed: %v\n", err)
	} else {
		if len(retrievedLargeValue) == len(largeValue) {
			match := true
			for i := range largeValue {
				if retrievedLargeValue[i] != largeValue[i] {
					match = false
					break
				}
			}
			if match {
				fmt.Printf("   ✓ GET (large) successful (took %v)\n", elapsed)
				fmt.Printf("   ✓ Reassembled %d bytes correctly\n", len(retrievedLargeValue))
			} else {
				fmt.Printf("   ✗ GET (large) value mismatch!\n")
			}
		} else {
			fmt.Printf("   ✗ GET (large) size mismatch!\n")
			fmt.Printf("   Expected: %d bytes\n", len(largeValue))
			fmt.Printf("   Got:      %d bytes\n", len(retrievedLargeValue))
		}
	}
	fmt.Println()

	// Test 6: GET (non-existent key)
	fmt.Println("6. Testing GET (non-existent key)...")
	start = time.Now()
	_, err = udpClient.Get("non-existent-key-go")
	elapsed = time.Since(start)
	if err != nil {
		if err.Error() == "key not found" {
			fmt.Printf("   ✓ GET (non-existent) correctly returned error (took %v)\n", elapsed)
			fmt.Printf("   ✓ Error: %v\n", err)
		} else {
			fmt.Printf("   ✗ GET (non-existent) returned unexpected error: %v\n", err)
		}
	} else {
		fmt.Printf("   ✗ GET (non-existent) should have failed but didn't!\n")
	}
	fmt.Println()

	// Test 7: Multiple concurrent SET/GET operations
	fmt.Println("7. Testing concurrent operations...")
	concurrentCount := 10
	successCount := 0
	errorCount := 0
	
	start = time.Now()
	done := make(chan bool, concurrentCount)
	for i := 0; i < concurrentCount; i++ {
		go func(id int) {
			key := fmt.Sprintf("concurrent-key-%d", id)
			value := []byte(fmt.Sprintf("concurrent-value-%d", id))
			
			if err := udpClient.Set(key, value, 3600); err != nil {
				errorCount++
				done <- false
				return
			}
			
			retrieved, err := udpClient.Get(key)
			if err != nil {
				errorCount++
				done <- false
				return
			}
			
			if string(retrieved) == string(value) {
				successCount++
				done <- true
			} else {
				errorCount++
				done <- false
			}
		}(i)
	}
	
	// Wait for all goroutines
	for i := 0; i < concurrentCount; i++ {
		<-done
	}
	elapsed = time.Since(start)
	
	fmt.Printf("   ✓ Concurrent operations completed (took %v)\n", elapsed)
	fmt.Printf("   ✓ Successful: %d/%d\n", successCount, concurrentCount)
	if errorCount > 0 {
		fmt.Printf("   ✗ Errors: %d\n", errorCount)
	}
	fmt.Println()

	fmt.Println("=== All Tests Complete ===")
	if successCount == concurrentCount && errorCount == 0 {
		fmt.Println("✓ All tests passed!")
		os.Exit(0)
	} else {
		fmt.Println("✗ Some tests failed")
		os.Exit(1)
	}
}

File: clients/java/src/main/java/com/blazecache/TestUDPCommands.java
package com.blazecache;

import java.io.IOException;
import java.util.Optional;

/**
 * Individual command tests for UDP client.
 */
public class TestUDPCommands {
    public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("Usage: TestUDPCommands <server-address>");
            System.err.println("Example: TestUDPCommands 127.0.0.1:6793");
            System.exit(1);
        }
        
        String serverAddr = args[0];
        
        try (UDPClient client = new UDPClient(serverAddr)) {
            System.out.println("╔════════════════════════════════════════════════════════════╗");
            System.out.println("║     Testing Java UDP Client - Individual Commands         ║");
            System.out.println("╚════════════════════════════════════════════════════════════╝");
            System.out.println("Server: " + serverAddr);
            System.out.println();
            
            int passed = 0;
            int failed = 0;
            
            // Test 1: PING
            System.out.println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
            System.out.println("Test 1: PING");
            System.out.println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
            try {
                client.ping();
                System.out.println("✓ PASS: PING successful");
                passed++;
            } catch (IOException e) {
                System.err.println("✗ FAIL: PING failed: " + e.getMessage());
                failed++;
            }
            System.out.println();
            
            // Test 2: PUT (small value)
            System.out.println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
            System.out.println("Test 2: PUT (small value)");
            System.out.println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
            try {
                String key = "test-key-small";
                String value = "Hello, BlazeCache UDP!";
                client.set(key, value.getBytes());
                System.out.println("✓ PASS: PUT successful");
                System.out.println("  Key: " + key);
                System.out.println("  Value: " + value);
                passed++;
            } catch (IOException e) {
                System.err.println("✗ FAIL: PUT failed: " + e.getMessage());
                failed++;
            }
            System.out.println();
            
            // Test 3: GET (small value)
            System.out.println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
            System.out.println("Test 3: GET (small value)");
            System.out.println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
            try {
                String key = "test-key-small";
                Optional<byte[]> result = client.get(key);
                if (result.isPresent()) {
                    String value = new String(result.get());
                    System.out.println("✓ PASS: GET successful");
                    System.out.println("  Key: " + key);
                    System.out.println("  Value: " + value);
                    if (value.equals("Hello, BlazeCache UDP!")) {
                        System.out.println("  ✓ Value matches expected");
                    } else {
                        System.out.println("  ✗ Value mismatch!");
                        failed++;
                    }
                    passed++;
                } else {
                    System.err.println("✗ FAIL: GET returned empty (key not found)");
                    failed++;
                }
            } catch (IOException e) {
                System.err.println("✗ FAIL: GET failed: " + e.getMessage());
                failed++;
            }
            System.out.println();
            
            // Test 4: GET (non-existent key)
            System.out.println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
            System.out.println("Test 4: GET (non-existent key)");
            System.out.println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
            try {
                String key = "non-existent-key-12345";
                Optional<byte[]> result = client.get(key);
                if (result.isEmpty()) {
                    System.out.println("✓ PASS: GET correctly returned empty for non-existent key");
                    passed++;
                } else {
                    System.err.println("✗ FAIL: GET should have returned empty");
                    failed++;
                }
            } catch (IOException e) {
                System.err.println("✗ FAIL: GET failed: " + e.getMessage());
                failed++;
            }
            System.out.println();
            
            // Test 5: PUT (large value - fragmentation)
            System.out.println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
            System.out.println("Test 5: PUT (large value - fragmentation)");
            System.out.println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
            try {
                String key = "test-key-large";
                byte[] largeValue = new byte[5000]; // Will be fragmented
                for (int i = 0; i < largeValue.length; i++) {
                    largeValue[i] = (byte) (i % 256);
                }
                client.set(key, largeValue);
                System.out.println("✓ PASS: PUT large message successful");
                System.out.println("  Key: " + key);
                System.out.println("  Size: " + largeValue.length + " bytes (fragmented)");
                passed++;
            } catch (IOException e) {
                System.err.println("✗ FAIL: PUT large message failed: " + e.getMessage());
                failed++;
            }
            System.out.println();
            
            // Test 6: GET (large value - reassembly)
            System.out.println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
            System.out.println("Test 6: GET (large value - reassembly)");
            System.out.println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
            try {
                String key = "test-key-large";
                Optional<byte[]> result = client.get(key);
                if (result.isPresent()) {
                    byte[] received = result.get();
                    System.out.println("✓ PASS: GET large message successful");
                    System.out.println("  Key: " + key);
                    System.out.println("  Size: " + received.length + " bytes (reassembled)");
                    if (received.length == 5000) {
                        System.out.println("  ✓ Size matches expected (5000 bytes)");
                        // Verify content
                        boolean matches = true;
                        for (int i = 0; i < received.length; i++) {
                            if (received[i] != (byte) (i % 256)) {
                                matches = false;
                                break;
                            }
                        }
                        if (matches) {
                            System.out.println("  ✓ Content matches expected");
                        } else {
                            System.out.println("  ✗ Content mismatch!");
                            failed++;
                        }
                    } else {
                        System.out.println("  ✗ Size mismatch! Expected 5000, got " + received.length);
                        failed++;
                    }
                    passed++;
                } else {
                    System.err.println("✗ FAIL: GET large message returned empty");
                    failed++;
                }
            } catch (IOException e) {
                System.err.println("✗ FAIL: GET large message failed: " + e.getMessage());
                failed++;
            }
            System.out.println();
            
            // Test 7: DELETE
            System.out.println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
            System.out.println("Test 7: DELETE");
            System.out.println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
            try {
                String key = "test-key-small";
                boolean deleted = client.delete(key);
                if (deleted) {
                    System.out.println("✓ PASS: DELETE successful");
                    System.out.println("  Key: " + key);
                    passed++;
                } else {
                    System.err.println("✗ FAIL: DELETE returned false (key not found)");
                    failed++;
                }
            } catch (IOException e) {
                System.err.println("✗ FAIL: DELETE failed: " + e.getMessage());
                failed++;
            }
            System.out.println();
            
            // Test 8: GET (after DELETE - should not exist)
            System.out.println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
            System.out.println("Test 8: GET (after DELETE - should not exist)");
            System.out.println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
            try {
                String key = "test-key-small";
                Optional<byte[]> result = client.get(key);
                if (result.isEmpty()) {
                    System.out.println("✓ PASS: GET correctly returned empty after DELETE");
                    passed++;
                } else {
                    System.err.println("✗ FAIL: GET should have returned empty after DELETE");
                    failed++;
                }
            } catch (IOException e) {
                System.err.println("✗ FAIL: GET failed: " + e.getMessage());
                failed++;
            }
            System.out.println();
            
            // Test 9: PUT with TTL
            System.out.println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
            System.out.println("Test 9: PUT with TTL");
            System.out.println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
            try {
                String key = "test-key-ttl";
                String value = "TTL test value";
                client.set(key, value.getBytes(), 3600); // 1 hour TTL
                System.out.println("✓ PASS: PUT with TTL successful");
                System.out.println("  Key: " + key);
                System.out.println("  Value: " + value);
                System.out.println("  TTL: 3600 seconds");
                passed++;
            } catch (IOException e) {
                System.err.println("✗ FAIL: PUT with TTL failed: " + e.getMessage());
                failed++;
            }
            System.out.println();
            
            // Summary
            System.out.println("╔════════════════════════════════════════════════════════════╗");
            System.out.println("║                      Test Summary                         ║");
            System.out.println("╚════════════════════════════════════════════════════════════╝");
            System.out.println("Total tests: " + (passed + failed));
            System.out.println("Passed: " + passed);
            System.out.println("Failed: " + failed);
            System.out.println();
            
            if (failed == 0) {
                System.out.println("✅ All tests passed!");
                System.exit(0);
            } else {
                System.out.println("❌ Some tests failed!");
                System.exit(1);
            }
            
        } catch (IOException e) {
            System.err.println("Failed to create UDP client: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }
}

